File: patterns/bayes/rules-python/bayes.py
from openhab import rule, Registry
from openhab.triggers import ItemStateChangeTrigger

PRIOR = 0.5
THRESHOLD = 0.85


def update_probability(prior, prob_given_true, prob_given_false):
    numerator = prob_given_true * prior
    return numerator / (numerator + prob_given_false * (1 - prior))


@rule(triggers=[
    ItemStateChangeTrigger("myObservation1"),
    ItemStateChangeTrigger("myObservation2"),
    ItemStateChangeTrigger("myObservation3"),
    ItemStateChangeTrigger("myObservation4"),
])
class BayesianSleepSensor:
    def execute(self, module, input):
        prob = PRIOR

        if str(Registry.getItem("myObservation1").getState()) == "OFF":
            prob = update_probability(prob, 0.99, 0.5)
        if str(Registry.getItem("myObservation2").getState()) == "NIGHT":
            prob = update_probability(prob, 0.75, 0.3)
        if str(Registry.getItem("myObservation3").getState()) == "OFF":
            prob = update_probability(prob, 0.9, 0.4)
        if str(Registry.getItem("myObservation4").getState()) == "ON":
            prob = update_probability(prob, 0.95, 0.5)

        Registry.getItem("mySleepProbability").postUpdate(prob * 100)
        Registry.getItem("mySleepSensor").sendCommand("ON" if prob >= THRESHOLD else "OFF")

        self.logger.info(
            "Schlafwahrscheinlichkeit: {:.1f}% (Schwelle: {:.0f}%)".format(prob * 100, THRESHOLD * 100)
        )

File: patterns/bayes/rules-js/bayes.js
const PRIOR = 0.5;
const THRESHOLD = 0.85;

function updateProbability(prior, probGivenTrue, probGivenFalse) {
  const numerator = probGivenTrue * prior;
  return numerator / (numerator + probGivenFalse * (1 - prior));
}

rules.JSRule({
  name: "Bayesian Sleep Sensor",
  triggers: [
    triggers.ItemStateChangeTrigger("myObservation1"),
    triggers.ItemStateChangeTrigger("myObservation2"),
    triggers.ItemStateChangeTrigger("myObservation3"),
    triggers.ItemStateChangeTrigger("myObservation4")
  ],
  execute: (event) => {
    let prob = PRIOR;

    if (items.getItem("myObservation1").state === "OFF") prob = updateProbability(prob, 0.99, 0.5);
    if (items.getItem("myObservation2").state === "NIGHT") prob = updateProbability(prob, 0.75, 0.3);
    if (items.getItem("myObservation3").state === "OFF") prob = updateProbability(prob, 0.9, 0.4);
    if (items.getItem("myObservation4").state === "ON") prob = updateProbability(prob, 0.95, 0.5);

    items.getItem("mySleepProbability").postUpdate(prob * 100);
    items.getItem("mySleepSensor").sendCommand(prob >= THRESHOLD ? "ON" : "OFF");
  }
});
